File: Algorithms/Recursion/permutation.c
#include <stdio.h>

void swap(char *x, char *y) {
    char temp = *x;
    *x = *y;
    *y = temp;
}

void permute(char *s, int l, int r) {
    if (l == r) {
        printf("%s\n", s);
    } else {
        for (int i = l; i <= r; i++) {
            swap((s + l), (s + i));  // Swap
            permute(s, l + 1, r);
            swap((s + l), (s + i));  // Backtrack
        }
    }
}

int main() {
    char str[] = "ABC";
    int n = 3;
    permute(str, 0, n - 1);
    return 0;
}

File: Algorithms/Recursion/permutation.py
def permute(s, l, r):
    if l == r:
        print(''.join(s))
    else:
        for i in range(l, r + 1):
            s[l], s[i] = s[i], s[l]  # Swap
            permute(s, l + 1, r)
            s[l], s[i] = s[i], s[l]  # Backtrack

# Example usage
string = "ABC"
n = len(string)
permute(list(string), 0, n - 1)
